Changes folder icons in hierarchy (Reflection approach)

garrafote opened the pull request:
I realize this has been discussed on #8 but I re-implemented this feature using reflections to change the actual folder icon eliminating the need of overdrawing the icon. As a result we get much better looking folders.

On the performance side, the icon injection is taking in average on my machine:

```
-   1 Folder:   0,07 ms
-  10 Folders:  0,16 ms
- 100 Folders:  0,99 ms
- 500 Folders:  4,50 ms
```

Subfolders inside collapsed folders have no impact on the numbers. This mean that if I have 10 collapsed folders the injection will take 0.16 ms independently of how many subfolders are inside these collapsed folders.

I believe these numbers can be crunched down even more with better reflection techniques.


One more thing: I also added a little fix to preserve sibling order when flattening folders.

mzage31 committed:
Changes folder icons in hierarchy

mzage31 committed:
Update HierarchyFolderIcon.cs

garrafote committed:
Merge branch 'master' of github.com:mzage31/unity-hierarchy-folders into icons

Commit:
Merge commit 'abd4a400c93429b198a91092a71e581162352c60' as 'Packages/UnityHierarchyFolders'

Commit:
Using reflection to set folder icons on the scene hierarchy

xsduan commented on Editor/HierarchyFolderIcon.cs:
```diff
@@ -0,0 +1,161 @@
+                data[index] = c;
+            }
+
+            var aaa = tinted.width * tinted.height * 4;
```
Could this probably have a better name? Unless AAA is some sort of technical term. Looks like it's calculating some sort of area?

garrafote replied:
terrible name, this variable represents the byte size of the current mipmap level on the texture.

xsduan commented on Runtime/Folder.cs:
```diff
@@ -56,6 +57,19 @@ public class Folder : MonoBehaviour
 
         private static Tool lastTool;
         private static Folder toolLock;
+        
+        public static HashSet<int> folders = new HashSet<int>();
+        //public static ReadOnlyCollection<int> folders => new ReadOnlyCollection<int>(_folders.ToList());
+
+        private void Start()
+        {
+            folders.Add(gameObject.GetInstanceID());
+        }
+
+        private void OnDestroy()
+        {
+            folders.Remove(gameObject.GetInstanceID());
+        }
```
What specifically is this intended to do?

garrafote replied:
This works as a cache of all the loaded folders. This cache helps to optimize the icon injection procedure as we only need to do know if a given instanceId is contained that HashSet to perform the icon injection.

I decided to have it on the actual Folder class to avoid the Folder class to depend on something else. This could certainly be refactored for better clarity.

xsduan commented on Editor/HierarchyFolderIcon.cs:
```diff
@@ -0,0 +1,161 @@
+                return;
+            }
+
+            _closedFolderTexture = (Texture2D) EditorGUIUtility.IconContent("Folder Icon").image;
+            _openFolderTexture = (Texture2D) EditorGUIUtility.IconContent("FolderEmpty Icon").image;
+ 
+            _openFolderSelectedTexture = GetTintedTexture(_openFolderTexture);
+            _closedFolderSelectedTexture = GetTintedTexture(_closedFolderTexture);
+            
+            const BindingFlags BindingAll = BindingFlags.Public 
+              | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.Instance;
+            
+            //public static List<SceneHierarchyWindow> GetAllSceneHierarchyWindows()
+            var assembly = typeof(SceneView).Assembly;
+
+            var sceneHierarchyWindowType = assembly.GetType("UnityEditor.SceneHierarchyWindow");
+            _getAllSceneHierarchyWindowsMethod = sceneHierarchyWindowType.GetMethod("GetAllSceneHierarchyWindows", BindingAll);
+            _sceneHierarchyProperty = sceneHierarchyWindowType.GetProperty("sceneHierarchy");
+
+            var sceneHierarchyType = assembly.GetType("UnityEditor.SceneHierarchy");
+            _treeViewProperty = sceneHierarchyType.GetProperty("treeView", BindingAll);
+            
+            var treeViewControllerType = assembly.GetType("UnityEditor.IMGUI.Controls.TreeViewController");
+            _dataProperty = treeViewControllerType.GetProperty("data", BindingAll);
+
+            var iTreeViewDataSourceType = assembly.GetType("UnityEditor.IMGUI.Controls.ITreeViewDataSource");
+            _getRowsMethod = iTreeViewDataSourceType.GetMethod("GetRows");
+            _isExpandedMethod = iTreeViewDataSourceType.GetMethod("IsExpanded", new Type[] {typeof(TreeViewItem)});
+            
+            var gameObjectTreeViewItemType = assembly.GetType("UnityEditor.GameObjectTreeViewItem");
+            _selectedIconProperty = gameObjectTreeViewItemType.GetProperty("selectedIcon", BindingAll);
+            _objectPPTRProperty = gameObjectTreeViewItemType.GetProperty("objectPPTR", BindingAll);
```
What are the guarantees of stability for these items? I don't want to spend too much time rehacking it every time there's a new upgrade if possible.

xsduan replied:
Additionally, can these properties be cached, or is there a concern that they may change?

garrafote replied:
Unfortunately we cant guarantee that this code wont break on future unity versions. But I've tested this code (not exactly this version, but another version of it with support for changing icon colors) on unity 2020.1.0b6 and it still worked.

The main idea is to get the current list of TreeViewItems being displayed in the scene hierarchy at the current frame. This list represent the visible items on the hierarchy at any given time. This could mean that some game object are culled if there are too many game objects on the scene. So, this list will change any time the hierarchy is scrolled, a game object is added or removed or a game object is collapsed/expanded. 

Another important detail is that there will be a new list of TreeViewItems for each active hierarchy window and this process needs to be repeated for each window.

xsduan reviewed (commented):
From what I can tell it's good, the rest is just non-functional cleanup that I'll do on my own time. However I do have a few questions about certain aspects. Mostly design concerns, no issues.

garrafote commented on Editor/HierarchyFolderIcon.cs:
```diff
@@ -0,0 +1,161 @@
+            {
+                tinted.SetPixelData(data, index, offset);
+                aaa >>= 2;
+                offset += aaa;
```
offset represents the starting index offset of the current mipmap on the texture.

garrafote commented:
On my private repository I've added support for folder icon colors. I tried to keep it simple and functional.

Compared to this pull request there is almost no difference on the performance. The main difference is that I generate the extra textures for the folder icons at startup time, I've added an int serialized property on the Folder component to hold the icon color index (that is stripped out on player builds), and I've changed the HashSet cache to an int, int dictionary that maps the instanceId to the the color index. On the injection procedure there is one more indirection level to get the colored icon and that is all. 

Just let me know if you are interested in having this feature on this repository I could push my color icon commits to this pull request together with the code revisions you mentioned above.

![image](https://user-images.githubusercontent.com/860128/80266433-3ada4c00-8672-11ea-8954-3d4cb97d3dc6.png)

xsduan commented:
It would certainly be something I’m very interested in having but I’d prefer if it was a separate pull request so that everything’s nicely separated. One step at a time and all that, I’ll certainly try to have this merged in by the end of today.
